src/Pinch/Protocol/Compact.hs

```diff
@@ -0,0 +1,487 @@
+    string (TE.encodeUtf8 $ messageName msg) <>
+    binarySerialize (messagePayload msg)
+
+binaryDeserializeMessage :: ByteString -> Either String Message
+binaryDeserializeMessage = runParser binaryMessageParser
+
+binaryMessageParser :: Parser Message
+binaryMessageParser = do
+    pid <- P.word8
+    when (pid /= protocolId) $ fail "Invalid protocol ID"
+    w <- P.word8
+    let ver = w .&. 0x1f
+    when (ver /= version) $ fail $ "Unsupported version: " ++ show ver
+    let code = w `shiftR` 5
+    msgId <- parseVarint
+    msgName <- TE.decodeUtf8 <$> (parseVarint >>= P.take . fromIntegral)
+    payload <- binaryParser ttype
+    return Message { messageType = case fromMessageCode code of
+                                     Nothing -> error $ "unknown message type: " ++ show code
+                                     Just t -> t
+                   , messageId = fromIntegral msgId
+                   , messageName = msgName
+                   , messagePayload = payload
+                   }
+
+
+------------------------------------------------------------------------------
+
+binaryDeserialize :: TType a -> ByteString -> Either String (ByteString, Value a)
+binaryDeserialize t = runParser' (binaryParser t)
+
+binaryParser :: TType a -> Parser (Value a)
+binaryParser typ = case typ of
+  TBool      -> do
+      n <- P.int8
+      return $ VBool (n == 1)
+  TByte      -> parseByte
+  TDouble    -> parseDouble
+  TInt16     -> parseInt16
+  TInt32     -> parseInt32
+  TInt64     -> parseInt64
+  TBinary    -> parseBinary
+  TStruct    -> parseStruct
+  TMap       -> parseMap
+  TSet       -> parseSet
+  TList      -> parseList
+
+intToZigZag :: Int64 -> Int64
+intToZigZag n =
+    (n `shiftL` 1) `xor` (n `shiftR` 63)
+
+zigZagToInt :: Int64 -> Int64
+zigZagToInt n =
+    fromIntegral (n' `shiftR` 1) `xor` (-(n .&. 1))
+  where
+    n' = fromIntegral n :: Word64
+    -- ensure no sign extension
+
+parseVarint :: Parser Int64
+parseVarint = go 0 0
+  where
+    go !val !shift = do
+        when (shift >= 64) $ fail "parseVarint: too wide"
+        n <- P.word8
+        let val' = val .|. ((fromIntegral n .&. 0x7f) `shiftL` shift)
+        if testBit n 7
+          then go val' (shift + 7)
+          else return val'
+
+getCType :: Word8 -> Parser SomeCType
+getCType code =
+    maybe (fail $ "Unknown CType: " ++ show code) return $ fromCompactCode code
+
+parseByte :: Parser (Value TByte)
+parseByte = VByte <$> P.int8
+
+parseDouble :: Parser (Value TDouble)
+parseDouble = VDouble <$> P.double
+
+parseInt16 :: Parser (Value TInt16)
+parseInt16 = VInt16 . fromIntegral . zigZagToInt <$> parseVarint
+
+parseInt32 :: Parser (Value TInt32)
+parseInt32 = VInt32 . fromIntegral . zigZagToInt <$> parseVarint
+
+parseInt64 :: Parser (Value TInt64)
+parseInt64 = VInt64 . fromIntegral . zigZagToInt <$> parseVarint
+
+parseBinary :: Parser (Value TBinary)
+parseBinary = do
+    n <- parseVarint
+    when (n < 0) $ fail "parseBinary: invalid length"
```

Comment:
Can you include the length in the error message? Should help with debugging if this is ever raised.